File: cpp_bind/reduced_softbody.cpp
#include "reduced_softbody.h"
#include "../src/BulletSoftBody/BulletReducedDeformableBody/btReducedDeformableBody.h"
#include "../src/BulletSoftBody/BulletReducedDeformableBody/btReducedDeformableBodyHelpers.h"
#include <vector>

nkReducedDeformableBodyHandle nk_reduced_softbody_create(
    nkSoftBodyWorldInfoHandle world_info,
    int node_count,
    const nkReal* positions,
    const nkReal* masses)
{
    if (!world_info || !positions || !masses || node_count <= 0) return nullptr;
    
    btSoftBodyWorldInfo* info = static_cast<btSoftBodyWorldInfo*>(world_info);
    
    std::vector<btVector3> x(node_count);
    std::vector<btScalar> m(node_count);
    
    for (int i = 0; i < node_count; i++) {
        x[i] = btVector3(positions[i * 3], positions[i * 3 + 1], positions[i * 3 + 2]);
        m[i] = masses[i];
    }
    
    btReducedDeformableBody* body = new btReducedDeformableBody(info, node_count, x.data(), m.data());
    return static_cast<nkReducedDeformableBodyHandle>(body);
}

void nk_reduced_softbody_destroy(nkReducedDeformableBodyHandle body)
{
    if (!body) return;
    btReducedDeformableBody* rsb = static_cast<btReducedDeformableBody*>(body);
    delete rsb;
}

void nk_reduced_softbody_set_reduced_modes(nkReducedDeformableBodyHandle body, int num_modes, int full_size)
{
    if (!body) return;
    btReducedDeformableBody* rsb = static_cast<btReducedDeformableBody*>(body);
    rsb->setReducedModes(num_modes, full_size);
}

int nk_reduced_softbody_get_num_reduced_modes(nkReducedDeformableBodyHandle body)
{
    if (!body) return 0;
    btReducedDeformableBody* rsb = static_cast<btReducedDeformableBody*>(body);
    return rsb->m_nReduced;
}

int nk_reduced_softbody_get_num_full_dofs(nkReducedDeformableBodyHandle body)
{
    if (!body) return 0;
    btReducedDeformableBody* rsb = static_cast<btReducedDeformableBody*>(body);
    return rsb->m_nFull;
}

void nk_reduced_softbody_set_stiffness_scale(nkReducedDeformableBodyHandle body, nkReal ks)
{
    if (!body) return;
    btReducedDeformableBody* rsb = static_cast<btReducedDeformableBody*>(body);
    rsb->setStiffnessScale(ks);
}

void nk_reduced_softbody_set_mass_scale(nkReducedDeformableBodyHandle body, nkReal rho)
{
    if (!body) return;
    btReducedDeformableBody* rsb = static_cast<btReducedDeformableBody*>(body);
    rsb->setMassScale(rho);
}

void nk_reduced_softbody_set_damping(nkReducedDeformableBodyHandle body, nkReal alpha, nkReal beta)
{
    if (!body) return;
    btReducedDeformableBody* rsb = static_cast<btReducedDeformableBody*>(body);
    rsb->setDamping(alpha, beta);
}

void nk_reduced_softbody_set_fixed_node(nkReducedDeformableBodyHandle body, int node_index)
{
    if (!body) return;
    btReducedDeformableBody* rsb = static_cast<btReducedDeformableBody*>(body);
    rsb->setFixedNodes(node_index);
}

void nk_reduced_softbody_disable_reduced_modes(nkReducedDeformableBodyHandle body, int rigid_only)
{
    if (!body) return;
    btReducedDeformableBody* rsb = static_cast<btReducedDeformableBody*>(body);
    rsb->disableReducedModes(rigid_only != 0);
}

void nk_reduced_softbody_set_rigid_velocity(nkReducedDeformableBodyHandle body, nkReal x, nkReal y, nkReal z)
{
    if (!body) return;
    btReducedDeformableBody* rsb = static_cast<btReducedDeformableBody*>(body);
    rsb->setRigidVelocity(btVector3(x, y, z));
}

void nk_reduced_softbody_set_rigid_angular_velocity(nkReducedDeformableBodyHandle body, nkReal x, nkReal y, nkReal z)
{
    if (!body) return;
    btReducedDeformableBody* rsb = static_cast<btReducedDeformableBody*>(body);
    rsb->setRigidAngularVelocity(btVector3(x, y, z));
}

void nk_reduced_softbody_get_rigid_velocity(nkReducedDeformableBodyHandle body, nkReal* out_x, nkReal* out_y, nkReal* out_z)
{
    if (!body) return;
    btReducedDeformableBody* rsb = static_cast<btReducedDeformableBody*>(body);
    const btVector3& v = rsb->getLinearVelocity();
    if (out_x) *out_x = v.x();
    if (out_y) *out_y = v.y();
    if (out_z) *out_z = v.z();
}

void nk_reduced_softbody_get_rigid_angular_velocity(nkReducedDeformableBodyHandle body, nkReal* out_x, nkReal* out_y, nkReal* out_z)
{
    if (!body) return;
    btReducedDeformableBody* rsb = static_cast<btReducedDeformableBody*>(body);
    const btVector3& w = rsb->getAngularVelocity();
    if (out_x) *out_x = w.x();
    if (out_y) *out_y = w.y();
    if (out_z) *out_z = w.z();
}

nkReal nk_reduced_softbody_get_total_mass(nkReducedDeformableBodyHandle body)
{
    if (!body) return 0.0f;
    btReducedDeformableBody* rsb = static_cast<btReducedDeformableBody*>(body);
    return rsb->getTotalMass();
}

void nk_reduced_softbody_set_total_mass(nkReducedDeformableBodyHandle body, nkReal mass)
{
    if (!body) return;
    btReducedDeformableBody* rsb = static_cast<btReducedDeformableBody*>(body);
    rsb->setTotalMass(mass);
}

void nk_reduced_softbody_get_rigid_transform(nkReducedDeformableBodyHandle body, nkTransform* out_transform)
{
    if (!body || !out_transform) return;
    btReducedDeformableBody* rsb = static_cast<btReducedDeformableBody*>(body);
    btTransform& t = rsb->getRigidTransform();
    btVector3 pos = t.getOrigin();
    btQuaternion rot = t.getRotation();
    out_transform->origin[0] = pos.x();
    out_transform->origin[1] = pos.y();
    out_transform->origin[2] = pos.z();
    out_transform->rotation[0] = rot.x();
    out_transform->rotation[1] = rot.y();
    out_transform->rotation[2] = rot.z();
    out_transform->rotation[3] = rot.w();
}

void nk_reduced_softbody_set_rigid_transform(nkReducedDeformableBodyHandle body, nkTransform* transform)
{
    if (!body || !transform) return;
    btReducedDeformableBody* rsb = static_cast<btReducedDeformableBody*>(body);
    btTransform t;
    t.setOrigin(btVector3(transform->origin[0], transform->origin[1], transform->origin[2]));
    t.setRotation(btQuaternion(transform->rotation[0], transform->rotation[1], transform->rotation[2], transform->rotation[3]));
    rsb->transformTo(t);
}

void nk_reduced_softbody_apply_central_impulse(nkReducedDeformableBodyHandle body, nkReal x, nkReal y, nkReal z)
{
    if (!body) return;
    btReducedDeformableBody* rsb = static_cast<btReducedDeformableBody*>(body);
    rsb->applyCentralImpulse(btVector3(x, y, z));
}

void nk_reduced_softbody_apply_torque_impulse(nkReducedDeformableBodyHandle body, nkReal x, nkReal y, nkReal z)
{
    if (!body) return;
    btReducedDeformableBody* rsb = static_cast<btReducedDeformableBody*>(body);
    rsb->applyTorqueImpulse(btVector3(x, y, z));
}

int nk_reduced_softbody_get_num_nodes(nkReducedDeformableBodyHandle body)
{
    if (!body) return 0;
    btReducedDeformableBody* rsb = static_cast<btReducedDeformableBody*>(body);
    return rsb->m_nodes.size();
}

void nk_reduced_softbody_get_node_position(nkReducedDeformableBodyHandle body, int index, nkReal* out_x, nkReal* out_y, nkReal* out_z)
{
    if (!body || index < 0) return;
    btReducedDeformableBody* rsb = static_cast<btReducedDeformableBody*>(body);
    if (index >= rsb->m_nodes.size()) return;
    
    const btVector3& pos = rsb->m_nodes[index].m_x;
    if (out_x) *out_x = pos.x();
    if (out_y) *out_y = pos.y();
    if (out_z) *out_z = pos.z();
}

void nk_reduced_softbody_get_node_rest_position(nkReducedDeformableBodyHandle body, int index, nkReal* out_x, nkReal* out_y, nkReal* out_z)
{
    if (!body || index < 0) return;
    btReducedDeformableBody* rsb = static_cast<btReducedDeformableBody*>(body);
    if (index >= (int)rsb->m_x0.size()) return;
    
    const btVector3& pos = rsb->m_x0[index];
    if (out_x) *out_x = pos.x();
    if (out_y) *out_y = pos.y();
    if (out_z) *out_z = pos.z();
}

nkReducedDeformableBodyHandle nk_reduced_softbody_create_from_vtk(
    nkSoftBodyWorldInfoHandle world_info,
    const char* vtk_file_path)
{
    if (!world_info || !vtk_file_path) return nullptr;
    btSoftBodyWorldInfo* info = static_cast<btSoftBodyWorldInfo*>(world_info);
    
    btReducedDeformableBody* body = btReducedDeformableBodyHelpers::createFromVtkFile(*info, vtk_file_path);
    return static_cast<nkReducedDeformableBodyHandle>(body);
}

void nk_reduced_softbody_read_reduced_info(nkReducedDeformableBodyHandle body, const char* file_path)
{
    if (!body || !file_path) return;
    btReducedDeformableBody* rsb = static_cast<btReducedDeformableBody*>(body);
    btReducedDeformableBodyHelpers::readReducedDeformableInfoFromFiles(rsb, file_path);
}

File: rust/src/core/reduced_softbody.rs
use std::ffi::{c_void, CString};
use std::ptr::NonNull;

use super::softbody::SoftBodyWorldInfo;
use super::types::{Real, Transform, Vec3};
use crate::ffi;

pub struct ReducedDeformableBody {
    handle: NonNull<c_void>,
}

impl ReducedDeformableBody {
    pub fn new(
        world_info: &SoftBodyWorldInfo,
        positions: &[[Real; 3]],
        masses: &[Real],
    ) -> Option<Self> {
        if positions.len() != masses.len() || positions.is_empty() {
            return None;
        }

        let handle = unsafe {
            ffi::nk_reduced_softbody_create(
                world_info.handle(),
                positions.len() as i32,
                positions.as_ptr() as *const Real,
                masses.as_ptr(),
            )
        };

        NonNull::new(handle).map(|h| Self { handle: h })
    }

    pub fn from_vtk(world_info: &SoftBodyWorldInfo, vtk_path: &str) -> Option<Self> {
        let c_path = CString::new(vtk_path).ok()?;
        let handle = unsafe {
            ffi::nk_reduced_softbody_create_from_vtk(world_info.handle(), c_path.as_ptr())
        };
        NonNull::new(handle).map(|h| Self { handle: h })
    }

    pub fn handle(&self) -> *mut c_void {
        self.handle.as_ptr()
    }

    pub fn set_reduced_modes(&self, num_modes: i32, full_size: i32) {
        unsafe {
            ffi::nk_reduced_softbody_set_reduced_modes(self.handle.as_ptr(), num_modes, full_size);
        }
    }

    pub fn num_reduced_modes(&self) -> i32 {
        unsafe { ffi::nk_reduced_softbody_get_num_reduced_modes(self.handle.as_ptr()) }
    }

    pub fn num_full_dofs(&self) -> i32 {
        unsafe { ffi::nk_reduced_softbody_get_num_full_dofs(self.handle.as_ptr()) }
    }

    pub fn set_stiffness_scale(&self, ks: Real) {
        unsafe {
            ffi::nk_reduced_softbody_set_stiffness_scale(self.handle.as_ptr(), ks);
        }
    }

    pub fn set_mass_scale(&self, rho: Real) {
        unsafe {
            ffi::nk_reduced_softbody_set_mass_scale(self.handle.as_ptr(), rho);
        }
    }

    pub fn set_damping(&self, alpha: Real, beta: Real) {
        unsafe {
            ffi::nk_reduced_softbody_set_damping(self.handle.as_ptr(), alpha, beta);
        }
    }

    pub fn set_fixed_node(&self, node_index: i32) {
        unsafe {
            ffi::nk_reduced_softbody_set_fixed_node(self.handle.as_ptr(), node_index);
        }
    }

    pub fn disable_reduced_modes(&self, rigid_only: bool) {
        unsafe {
            ffi::nk_reduced_softbody_disable_reduced_modes(
                self.handle.as_ptr(),
                if rigid_only { 1 } else { 0 },
            );
        }
    }

    pub fn set_rigid_velocity(&self, v: Vec3) {
        unsafe {
            ffi::nk_reduced_softbody_set_rigid_velocity(self.handle.as_ptr(), v.x, v.y, v.z);
        }
    }

    pub fn set_rigid_angular_velocity(&self, w: Vec3) {
        unsafe {
            ffi::nk_reduced_softbody_set_rigid_angular_velocity(self.handle.as_ptr(), w.x, w.y, w.z);
        }
    }

    pub fn rigid_velocity(&self) -> Vec3 {
        let mut x: Real = 0.0;
        let mut y: Real = 0.0;
        let mut z: Real = 0.0;
        unsafe {
            ffi::nk_reduced_softbody_get_rigid_velocity(self.handle.as_ptr(), &mut x, &mut y, &mut z);
        }
        Vec3::new(x, y, z)
    }

    pub fn rigid_angular_velocity(&self) -> Vec3 {
        let mut x: Real = 0.0;
        let mut y: Real = 0.0;
        let mut z: Real = 0.0;
        unsafe {
            ffi::nk_reduced_softbody_get_rigid_angular_velocity(self.handle.as_ptr(), &mut x, &mut y, &mut z);
        }
        Vec3::new(x, y, z)
    }

    pub fn total_mass(&self) -> Real {
        unsafe { ffi::nk_reduced_softbody_get_total_mass(self.handle.as_ptr()) }
    }

    pub fn set_total_mass(&self, mass: Real) {
        unsafe {
            ffi::nk_reduced_softbody_set_total_mass(self.handle.as_ptr(), mass);
        }
    }

    pub fn rigid_transform(&self) -> Transform {
        let mut transform: ffi::nkTransform = unsafe { std::mem::zeroed() };
        unsafe {
            ffi::nk_reduced_softbody_get_rigid_transform(self.handle.as_ptr(), &mut transform);
        }
        Transform {
            position: Vec3::new(transform.origin.data[0], transform.origin.data[1], transform.origin.data[2]),
            rotation: glam::Quat::from_xyzw(
                transform.rotation.data[0],
                transform.rotation.data[1],
                transform.rotation.data[2],
                transform.rotation.data[3],
            ),
        }
    }

    pub fn set_rigid_transform(&self, transform: &Transform) {
        let nk_transform = ffi::nkTransform {
            origin: ffi::nkVector3 {
                data: [transform.position.x, transform.position.y, transform.position.z],
            },
            rotation: ffi::nkQuaternion {
                data: [transform.rotation.x, transform.rotation.y, transform.rotation.z, transform.rotation.w],
            },
        };
        unsafe {
            ffi::nk_reduced_softbody_set_rigid_transform(self.handle.as_ptr(), &nk_transform as *const _ as *mut _);
        }
    }

    pub fn apply_central_impulse(&self, impulse: Vec3) {
        unsafe {
            ffi::nk_reduced_softbody_apply_central_impulse(self.handle.as_ptr(), impulse.x, impulse.y, impulse.z);
        }
    }

    pub fn apply_torque_impulse(&self, torque: Vec3) {
        unsafe {
            ffi::nk_reduced_softbody_apply_torque_impulse(self.handle.as_ptr(), torque.x, torque.y, torque.z);
        }
    }

    pub fn num_nodes(&self) -> i32 {
        unsafe { ffi::nk_reduced_softbody_get_num_nodes(self.handle.as_ptr()) }
    }

    pub fn node_position(&self, index: i32) -> Option<Vec3> {
        if index < 0 || index >= self.num_nodes() {
            return None;
        }
        let mut x: Real = 0.0;
        let mut y: Real = 0.0;
        let mut z: Real = 0.0;
        unsafe {
            ffi::nk_reduced_softbody_get_node_position(self.handle.as_ptr(), index, &mut x, &mut y, &mut z);
        }
        Some(Vec3::new(x, y, z))
    }

    pub fn node_rest_position(&self, index: i32) -> Option<Vec3> {
        if index < 0 || index >= self.num_nodes() {
            return None;
        }
        let mut x: Real = 0.0;
        let mut y: Real = 0.0;
        let mut z: Real = 0.0;
        unsafe {
            ffi::nk_reduced_softbody_get_node_rest_position(self.handle.as_ptr(), index, &mut x, &mut y, &mut z);
        }
        Some(Vec3::new(x, y, z))
    }

    pub fn read_reduced_info(&self, file_path: &str) {
        if let Ok(c_path) = CString::new(file_path) {
            unsafe {
                ffi::nk_reduced_softbody_read_reduced_info(self.handle.as_ptr(), c_path.as_ptr());
            }
        }
    }
}

impl Drop for ReducedDeformableBody {
    fn drop(&mut self) {
        unsafe {
            ffi::nk_reduced_softbody_destroy(self.handle.as_ptr());
        }
    }
}

unsafe impl Send for ReducedDeformableBody {}
unsafe impl Sync for ReducedDeformableBody {}

#[cfg(test)]
mod tests_reduced_softbody {
    use super::*;
    use crate::core::world::PhysicsWorldBuilder;

    fn create_world() -> crate::core::world::PhysicsWorld {
        PhysicsWorldBuilder::new().build()
    }

    #[test]
    fn test_reduced_softbody_create() {
        let world = create_world();
        let info = world.softbody_world_info();
        
        let positions = [
            [0.0, 0.0, 0.0],
            [1.0, 0.0, 0.0],
            [0.0, 1.0, 0.0],
            [0.0, 0.0, 1.0],
        ];
        let masses = [1.0, 1.0, 1.0, 1.0];
        
        let body = ReducedDeformableBody::new(&info, &positions, &masses);
        assert!(body.is_some());
        
        let body = body.unwrap();
        assert_eq!(body.num_nodes(), 4);
    }

    #[test]
    fn test_reduced_softbody_modes() {
        let world = create_world();
        let info = world.softbody_world_info();
        
        let positions = [[0.0, 0.0, 0.0]; 4];
        let masses = [1.0; 4];
        
        let body = ReducedDeformableBody::new(&info, &positions, &masses).unwrap();
        
        body.set_reduced_modes(3, 12);
        assert_eq!(body.num_reduced_modes(), 3);
        assert_eq!(body.num_full_dofs(), 12);
    }

    #[test]
    fn test_reduced_softbody_velocity() {
        let world = create_world();
        let info = world.softbody_world_info();
        
        let positions = [[0.0, 0.0, 0.0]; 4];
        let masses = [1.0; 4];
        
        let body = ReducedDeformableBody::new(&info, &positions, &masses).unwrap();
        
        body.set_rigid_velocity(Vec3::new(1.0, 2.0, 3.0));
        let v = body.rigid_velocity();
        assert!((v.x - 1.0).abs() < 0.001);
        assert!((v.y - 2.0).abs() < 0.001);
        assert!((v.z - 3.0).abs() < 0.001);
    }

    #[test]
    fn test_reduced_softbody_mass() {
        let world = create_world();
        let info = world.softbody_world_info();
        
        let positions = [[0.0, 0.0, 0.0]; 4];
        let masses = [1.0; 4];
        
        let body = ReducedDeformableBody::new(&info, &positions, &masses).unwrap();
        
        body.set_total_mass(10.0);
        assert!((body.total_mass() - 10.0).abs() < 0.001);
    }
}
